### ISSUE TITLE
[vtorc] PostMasterFailoverProcesses callback invalid

### ISSUE THREAD

**Issue by vczyh** [vitessio/vitess#8769]
#### Overview

vtorc `config.json` `PostMasterFailoverProcesses ` invalid.

#### Reproduction Steps

Steps to reproduce this issue, example:

1. Deploy `vtorc`

```bash
vtorc \
  -topo_implementation etcd2 -topo_global_server_address 10.0.44.96:2379,10.0.44.97:2379,10.0.44.98:2379 -topo_global_root /vitess/global \
  -log_dir /vitess/log/vtorc \
  -config "/vitess/component/vtorc/orc_config.json" \
  -orc_web_dir "/vitess/component/vtorc/orchestrator" \
  -clusters_to_watch default-ks
```

```json
{
  "Debug": true,
  "MySQLTopologyUser": "orc_client_user",
  "MySQLTopologyPassword": "MySQL_321@T6",
  "MySQLReplicaUser": "vt_repl",
  "MySQLReplicaPassword": "MySQL_321@T6",
  "RecoveryPeriodBlockSeconds": 5,
  "PostMasterFailoverProcesses": [
    "echo 'Recovered from {failureType} on {failureCluster}. Failed: {failedHost}:{failedPort}; Promoted: {successorHost}:{successorPort}' >> /tmp/recovery.log"
  ]
}
```

2. Deploy one shard. Then `vtorc` automatically init `master` 

3. Stop master `tablet`. `vtorc` failover successfully.

4. But `PostMasterFailoverProcesses` invalid. nothing in `/tmp/recovery.log`.

#### Operating system and Environment details

- CentOS Linux release 8.2.2004 (Linux 4.18.0-193.el8.x86_64)
- [Vitess v10.0.2](https://github.com/vitessio/vitess/releases/tag/v10.0.2)

#### Log Fragments

vtorc log

```log
I0902 09:23:53.978147  113726 locks.go:355] Locking shard default-ks/0 for action Orc Recovery
I0902 09:23:54.155914  113726 locks.go:395] Unlocking shard default-ks/0 for successful action Orc Recovery
I0902 09:37:24.010585  113726 locks.go:355] Locking shard default-ks/0 for action Orc Recovery
I0902 09:37:24.366794  113726 locks.go:395] Unlocking shard default-ks/0 for successful action Orc Recovery
I0902 09:37:24.582382  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:25.019331  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:25.797494  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:26.171078  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:26.252008  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:26.410617  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:26.632884  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:26.829573  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:27.029573  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
I0902 09:37:27.349748  113726 log.go:255] Unable to determine cluster name. clusterHint=10.0.44.25:3306
```

`journalctl  -u vtorc`
import line i think: `ERROR topology_recovery.go:1773] tablet alias is nil`
```log
9月 02 15:38:07 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:07     INFO tablet_discovery.go:219] Forgetting: alias:<cell:"cell1" uid:2 > hostname:"10.0.44.26" port_map:<key:"grpc" value:17100 > port_map:<key:"vt" value:17000 > keyspace:"default-ks" shard:"0" type:MASTER mysql_hostname:"10.0.44.26" mysql_port:3306 master_term_start_time:<seconds:1630546644 nanoseconds:191729544 >
9月 02 15:38:07 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:07     INFO audit_dao.go:105] auditType:forget instance:10.0.44.26:3306 cluster: message:
9月 02 15:38:08 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:08     INFO topology_recovery.go:1652] executeCheckAndRecoverFunction: proceeding with ClusterHasNoMaster detection on 10.0.44.25:3306; isActionable?: true; skipProcesses: false
9月 02 15:38:08 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:08     INFO topology_recovery.go:1531] topology_recovery: detected ClusterHasNoMaster failure on 10.0.44.25:3306
9月 02 15:38:08 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:08     INFO topology_recovery.go:229] topology_recovery: No OnFailureDetectionProcesses hooks to run
9月 02 15:38:08 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:08     INFO topology_recovery.go:1704] executeCheckAndRecoverFunction: proceeding with ClusterHasNoMaster recovery on 10.0.44.25:3306; isRecoverable?: true; skipProcesses: false
9月 02 15:38:08 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:08     INFO topology_recovery.go:2054] Analysis: ClusterHasNoMaster, will elect a new master: default-ks:0
9月 02 15:38:08 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:08    DEBUG instance_topology_dao.go:346] Stopping 2 replicas via StopReplicationNice
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:284] Stopped replication nicely on 10.0.44.27:3306, Self:mysql-bin.000002:11926, Exec:mysql-bin.000002:11786
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:284] Stopped replication nicely on 10.0.44.25:3306, Self:mysql-bin.000002:11884, Exec:mysql-bin.000002:11786
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:394] Stopped replication on 10.0.44.27:3306, Self:mysql-bin.000002:11926, Exec:mysql-bin.000002:11786
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:394] Stopped replication on 10.0.44.25:3306, Self:mysql-bin.000002:11884, Exec:mysql-bin.000002:11786
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology.go:625] Will move 10.0.44.25:3306 below 10.0.44.27:3306 via GTID
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO audit_dao.go:105] auditType:begin-maintenance instance:10.0.44.25:3306 cluster:10.0.44.25:3306 message:maintenanceToken: 10, owner: etcd-01, reason: move below 10.0.44.27:3306
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:394] Stopped replication on 10.0.44.25:3306, Self:mysql-bin.000002:11884, Exec:mysql-bin.000002:11786
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09    DEBUG instance_topology_dao.go:597] ChangeMasterTo: will attempt changing master on 10.0.44.25:3306 to 10.0.44.27:3306, mysql-bin.000002:11926
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:701] ChangeMasterTo: Changed master on 10.0.44.25:3306 to: 10.0.44.27:3306, mysql-bin.000002:11926. GTID: true
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:436] Started replication on 10.0.44.25:3306
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO audit_dao.go:105] auditType:move-below-gtid instance:10.0.44.25:3306 cluster:10.0.44.25:3306 message:moved 10.0.44.25:3306 below 10.0.44.27:3306
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO audit_dao.go:105] auditType:end-maintenance instance:10.0.44.25:3306 cluster:10.0.44.25:3306 message:maintenanceToken: 10
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:229] topology_recovery: - electNewMaster: applying semi-sync false: success=true
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology_dao.go:939] instance 10.0.44.27:3306 read_only: false
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO audit_dao.go:105] auditType:read-only instance:10.0.44.27:3306 cluster:10.0.44.27:3306 message:set as false
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:229] topology_recovery: - electNewMaster: set read-only false: success=true
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:1714] Topology recovery: {"Id":13,"UID":"1630568288973939722:960f4e1c09c59bfd8f6c00ead9326d80a3af22ca258658b7c476c54fe3f478ac","AnalysisEntry":{"AnalyzedInstanceKey":{"Hostname":"10.0.44.25","Port":3306},"AnalyzedInstanceMasterKey":{"Hostname":"10.0.44.26","Port":3306},"TabletType":2,"MasterTimeStamp":"0001-01-01T00:00:00Z","SuggestedClusterAlias":"default-ks:0","ClusterDetails":{"ClusterName":"10.0.44.26:3306","ClusterAlias":"default-ks:0","ClusterDomain":"10.0.44.26:3306","CountInstances":0,"HeuristicLag":0,"HasAutomatedMasterRecovery":true,"HasAutomatedIntermediateMasterRecovery":false},"AnalyzedInstanceDataCenter":"cell1","AnalyzedInstanceRegion":"","AnalyzedInstancePhysicalEnvironment":"","AnalyzedInstanceBinlogCoordinates":{"LogFile":"mysql-bin.000002","LogPos":11884,"Type":0},"IsMaster":false,"IsClusterMaster":false,"IsCoMaster":false,"LastCheckValid":true,"LastCheckPartialSuccess":true,"CountReplicas":0,"CountValidReplicas":0,"CountValidReplicatingReplicas":0,"CountReplicasFailingToConnectToMaster":0,"CountDowntimedReplicas":0,"ReplicationDepth":1,"Replicas":[],"SlaveHosts":[],"IsFailingToConnectToMaster":false,"ReplicationStopped":false,"Analysis":"ClusterHasNoMaster","Description":"Cluster has no master","StructureAnalysis":null,"IsDowntimed":false,"IsReplicasDowntimed":false,"DowntimeEndTimestamp":"","DowntimeRemainingSeconds":0,"IsBinlogServer":false,"PseudoGTIDImmediateTopology":false,"OracleGTIDImmediateTopology":false,"MariaDBGTIDImmediateTopology":false,"BinlogServerImmediateTopology":false,"SemiSyncMasterEnabled":false,"SemiSyncMasterStatus":false,"SemiSyncMasterWaitForReplicaCount":1,"SemiSyncMasterClients":0,"SemiSyncReplicaEnabled":false,"CountSemiSyncReplicasEnabled":0,"CountLoggingReplicas":0,"CountStatementBasedLoggingReplicas":0,"CountMixedBasedLoggingReplicas":0,"CountRowBasedLoggingReplicas":0,"CountDistinctMajorVersionsLoggingReplicas":0,"CountDelayedReplicas":0,"CountLaggingReplicas":0,"IsActionableRecovery":true,"ProcessingNodeHostname":"etcd-01","ProcessingNodeToken":"a3233ae9886cc783c1f2fb16d0f8597344869f5a9d2279e3cb5d40dc65da6feb","CountAdditionalAgreeingNodes":0,"StartActivePeriod":"","SkippableDueToDowntime":false,"GTIDMode":"ON","MinReplicaGTIDMode":"","MaxReplicaGTIDMode":"","MaxReplicaGTIDErrant":"","CommandHint":"","IsReadOnly":true},"SuccessorKey":null,"SuccessorAlias":"","IsActive":false,"IsSuccessful":false,"LostReplicas":[],"ParticipatingInstanceKeys":[],"AllErrors":[],"RecoveryStartTimestamp":"","RecoveryEndTimestamp":"","ProcessingNodeHostname":"","ProcessingNodeToken":"","Acknowledged":false,"AcknowledgedAt":"","AcknowledgedBy":"","AcknowledgedComment":"","LastDetectionId":0,"RelatedRecoveryId":0,"Type":"","RecoveryType":"NotMasterRecovery"}
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:229] topology_recovery: No PostUnsuccessfulFailoverProcesses hooks to run
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:229] topology_recovery: Waiting for 0 postponed functions
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09    DEBUG postponed_functions.go:52] PostponedFunctionsContainer: waiting on 0 postponed functions
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09    DEBUG postponed_functions.go:54] PostponedFunctionsContainer: done waiting
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:229] topology_recovery: Executed 0 postponed functions
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:1652] executeCheckAndRecoverFunction: proceeding with MasterHasMaster detection on 10.0.44.27:3306; isActionable?: true; skipProcesses: false
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:1531] topology_recovery: detected MasterHasMaster failure on 10.0.44.27:3306
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:229] topology_recovery: No OnFailureDetectionProcesses hooks to run
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:1704] executeCheckAndRecoverFunction: proceeding with MasterHasMaster recovery on 10.0.44.27:3306; isRecoverable?: true; skipProcesses: false
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO topology_recovery.go:2143] Analysis: MasterHasMaster, will fix incorrect mastership 10.0.44.27:3306
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO instance_topology.go:1045] Will reset replica on 10.0.44.27:3306
9月 02 15:38:09 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:09     INFO audit_dao.go:105] auditType:begin-maintenance instance:10.0.44.27:3306 cluster:10.0.44.27:3306 message:maintenanceToken: 11, owner: etcd-01, reason: reset replica
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO instance_topology_dao.go:394] Stopped replication on 10.0.44.27:3306, Self:mysql-bin.000002:11926, Exec:mysql-bin.000002:11786
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO instance_topology_dao.go:763] Reset replication 10.0.44.27:3306
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO audit_dao.go:105] auditType:reset-slave instance:10.0.44.27:3306 cluster:10.0.44.27:3306 message:10.0.44.27:3306 replication reset
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO audit_dao.go:105] auditType:end-maintenance instance:10.0.44.27:3306 cluster:10.0.44.27:3306 message:maintenanceToken: 11
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO topology_recovery.go:1652] executeCheckAndRecoverFunction: proceeding with DeadMaster detection on 10.0.44.26:3306; isActionable?: true; skipProcesses: false
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO topology_recovery.go:1531] topology_recovery: detected DeadMaster failure on 10.0.44.26:3306
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO topology_recovery.go:229] topology_recovery: No OnFailureDetectionProcesses hooks to run
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO topology_recovery.go:1704] executeCheckAndRecoverFunction: proceeding with DeadMaster recovery on 10.0.44.26:3306; isRecoverable?: true; skipProcesses: false
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO topology_recovery.go:838] Analysis: DeadMaster, deadmaster 10.0.44.26:3306
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10     INFO topology_recovery.go:845] CheckAndRecover: Analysis: DeadMaster, InstanceKey: 10.0.44.26:3306, candidateInstanceKey: 10.0.44.27:3306, skipProcesses: false: NOT detecting/recovering host, could not obtain shard lock (tablet alias is nil)
9月 02 15:38:10 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:10    ERROR topology_recovery.go:1773] tablet alias is nil
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37    DEBUG orchestrator.go:396] kv.SubmitMastersToKvStores, clusterName: , force: false: numPairs: 5
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37    DEBUG orchestrator.go:419] kv.SubmitMastersToKvStores: submitKvPairs: 0
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37     INFO audit_dao.go:105] auditType:review-unseen-instances instance::0 cluster: message:Operations: 0
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37     INFO audit_dao.go:105] auditType:inject-unseen-masters instance::0 cluster: message:Operations: 0
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37     INFO audit_dao.go:105] auditType:forget-unseen instance::0 cluster: message:Forgotten instances: 0
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37     INFO audit_dao.go:105] auditType:forget-unseen-differently-resolved instance::0 cluster: message:Forgotten instances: 0
9月 02 15:38:37 etcd-01 vtorc_start.sh[113723]: 2021-09-02 15:38:37     INFO audit_dao.go:105] auditType:resolve-unknown-masters instance::0 cluster: message:Num resolved hostnames: 0 
```

**Comment by GuptaManan100** [comment]
Closing since there has been no activity in over a year. Please open a new issue with updated information if the problem persists.
